relay-server/src/actors/project.rs

```diff
@@ -557,11 +587,35 @@ impl Project {
                     .as_ref()
                     .map(|resp| resp.is_local)
                     .unwrap_or(false);
-                slf.state = state_result.map(|resp| resp.state).ok();
-
-                if let Some(ref state) = slf.state {
-                    log::debug!("project {} state updated", id);
-                    sender.send(state.clone()).ok();
+                match state_result {
+                    Ok(response) => {
+                        let send_state = if let Some(ref mut saved_state) = slf.state {
+                            if *saved_state != response.state {
+                                log::debug!("project {} state did change, updated", id);
+                                *saved_state = response.state.clone();
+                            } else {
+                                // Avoid assigning the newly parsed state if it didn't actually change.
+                                // We have a few lazycells and internal caches of regexes in
+                                // projectconfigs (related to PII processing) that would otherwise be
+                                // unnecessarily wiped.
+                                log::debug!("project {} state did not change", id);
+                            }
+
+                            saved_state.clone()
+                        } else {
+                            log::debug!("project {} state created", id);
+                            slf.state = Some(response.state.clone());
+                            response.state
+                        };
+
+                        slf.last_fetch = Utc::now();
+                        sender.send(send_state).ok();
+                    }
+                    Err(()) => {
+                        // should be unreachable
+                        log::error!("received unit-error from FetchProjectState");
```

Comment:
This is not unreachable, and the error condition is already logged.

Reply:
Or not, looks like we might swallow the errors in some cases.

Reply:
The one I was going off of was https://github.com/getsentry/relay/blob/b58ce9e707e92e1ff8540e90ed8c911452afb1c5/relay-server/src/actors/project_cache.rs#L212... so a few mailbox errors bubble up like this. It's not a very descriptive error message but I would split it up once it happens